src/libraries/System.Security.Cryptography/tests/X509Certificates/DynamicChainTests.cs

```diff
@@ -594,6 +596,102 @@ public static void NameConstraintViolation_InvalidGeneralNames()
+                        {
+                            OtherName = new OtherNameAsn
+                            {
+                                TypeId = "1.3.6.1.4.1.311.20.2.3", //User Principal Name (UPN)
+                                Value = github,
+                            }
+                        }
+                    }
+                ]
+            };
+
+            nameConstraints.Encode(writer);
+            string encoded = writer.Encode(Convert.ToHexString);
+
+            TestNameConstrainedChain(encoded, builder, (bool result, X509Chain chain) => {
+                Assert.False(result, "chain.Build");
+
+                if (PlatformDetection.IsWindows)
+                {
+                    Assert.Equal(X509ChainStatusFlags.HasExcludedNameConstraint, chain.AllStatusFlags());
+                }
+                else
+                {
+                    Assert.Equal(
+                        PlatformNameConstraints(X509ChainStatusFlags.HasNotSupportedNameConstraint, true),
+                        chain.AllStatusFlags());
+                }
+            });
+        }
+
+        [Fact]
+        public static void NameConstraintViolation_PermittedTree_Upn()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+            builder.AddUserPrincipalName("mona@github.com");
+
+            AsnWriter writer = new(AsnEncodingRules.DER);
+            writer.WriteCharacterString(UniversalTagNumber.UTF8String, "@microsoft.com");
+            byte[] github = writer.Encode();
+            writer.Reset();
+
+            NameConstraintsAsn nameConstraints = new NameConstraintsAsn
+            {
+                permittedSubtrees =
+                [
+                    new GeneralSubtreeAsn
+                    {
+                        @base = new GeneralNameAsn
+                        {
+                            OtherName = new OtherNameAsn
+                            {
+                                TypeId = "1.3.6.1.4.1.311.20.2.3", //User Principal Name (UPN)
+                                Value = github,
+                            }
+                        }
+                    }
+                ]
+            };
+
+            nameConstraints.Encode(writer);
+            string encoded = writer.Encode(Convert.ToHexString);
+
+            TestNameConstrainedChain(encoded, builder, (bool result, X509Chain chain) => {
+                Assert.False(result, "chain.Build");
+
+                if (PlatformDetection.IsWindows)
+                {
+                    Assert.Equal(X509ChainStatusFlags.HasNotPermittedNameConstraint, chain.AllStatusFlags());
+                }
+                else
+                {
+                    Assert.Equal(
+                        PlatformNameConstraints(X509ChainStatusFlags.HasNotSupportedNameConstraint, true),
+                        chain.AllStatusFlags());
+                }
+            });
+        }
+
```

Comment:
Consider adding tests with a UPN constraint that passes (on Windows), so we're set up for a future where we gain platform support (or write it ourselves).

Reply:
Yeah, incomplete tests was one of the reason I put a "NO REVIEW" label on the issue :-)